tests/integration/ha/test_ha.py

```diff
@@ -37,8 +75,70 @@ async def test_build_and_deploy(ops_test: OpsTest, kafka_charm):
         assert ops_test.model.applications[APP_NAME].status == "active"
         assert ops_test.model.applications[ZK_NAME].status == "active"
 
+    await ops_test.model.add_relation(APP_NAME, f"{DUMMY_NAME}:{REL_NAME_ADMIN}")
+    await ops_test.model.wait_for_idle(apps=[APP_NAME, DUMMY_NAME], idle_period=30)
+    assert ops_test.model.applications[APP_NAME].status == "active"
+    assert ops_test.model.applications[DUMMY_NAME].status == "active"
+
+
+async def test_replicated_events(ops_test: OpsTest):
+    await ops_test.model.applications[APP_NAME].add_units(count=2)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", timeout=600, idle_period=120, wait_for_exact_units=3
+    )
+    logger.info("Producing messages and checking on all units")
+    produce_and_check_logs(
+        model_full_name=ops_test.model_full_name,
+        kafka_unit_name=f"{APP_NAME}/0",
+        provider_unit_name=f"{DUMMY_NAME}/0",
+        topic="replicated-topic",
+        replication_factor=3,
+        num_partitions=1,
+    )
+    # check offsets in the two remaining units
+    assert await get_topic_offsets(
+        ops_test=ops_test, topic="replicated-topic", unit_name=f"{APP_NAME}/0"
+    ) == ["0", "15"]
+
+
+async def test_kill_broker_with_topic_leader(
+    ops_test: OpsTest,
+    c_writes: ContinuousWrites,
+    c_writes_runner: ContinuousWrites,
+    restart_delay,
+):
+    initial_leader_num = await get_topic_leader(
+        ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME
+    )
+    initial_offsets = await get_topic_offsets(
+        ops_test=ops_test,
+        topic=ContinuousWrites.TOPIC_NAME,
+        unit_name=f"kafka/{initial_leader_num}",
+    )
+
+    logger.info(
+        f"Killing broker of leader for topic '{ContinuousWrites.TOPIC_NAME}': {initial_leader_num}"
+    )
+    await send_control_signal(
+        ops_test=ops_test, unit_name=f"{APP_NAME}/{initial_leader_num}", kill_code="SIGKILL"
+    )
+    # Give time for the service to restart
+    time.sleep(10)
+
+    # Check that leader changed
+    next_leader_num = await get_topic_leader(ops_test=ops_test, topic="replicated-topic")
+    next_offsets = await get_topic_offsets(
+        ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME, unit_name=f"kafka/{next_leader_num}"
+    )
+    res = c_writes.stop()
+
+    assert initial_leader_num != next_leader_num
+    assert int(next_offsets[-1]) > int(initial_offsets[-1])
+    assert res.lost_messages == "0"
+    assert res.count == int(res.last_expected_message)
```

Comment:
nit: this whole continuous writes consistency logic could be put in a helper function so all tests are consistent by being systematically added to the end each test
(counts, testing the replication on all the nodes etc...)

Reply:
yes, I'm changing this on next PRs :)